src/formats/base_32.cpp

```diff
@@ -0,0 +1,187 @@
+    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1
+};
+const static char separator = '1';
+const static size_t checksum_length = 6;
+const static size_t hr_part_min_length = 1;
+const static size_t hr_part_max_length = 83;
+const static size_t data_part_min_length = checksum_length;
+const static size_t base32_min_length = hr_part_min_length + 1 + data_part_min_length;
+const static size_t base32_max_length = 90;
+const static size_t data_part_max_length = base32_max_length - hr_part_min_length - 1;
+
+uint32_t polymod_base32(const data_chunk& values)
+{
+    static const std::vector<uint32_t> gen = 
+        {0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3};
+    uint32_t chk = 1;
+    for (auto v : values)
+    {
+        const uint32_t b = (chk >> 25);
+        chk = (chk & 0x1ffffff) << 5 ^ v;
+        for (auto i = 0; i < 5; ++i)
+        {
+            chk ^= ((b >> i) & 1) ? gen[i] : 0;
+        }
+    }
+    return chk;
+}
+
+data_chunk hrp_expand_base32(const std::string& s)
+{
+    data_chunk result(2 * s.size() + 1);
+    auto iterator = result.begin();
+    for (size_t i = 0; i < s.size(); ++i)
+    {
+        *iterator = s[i] >> 5;
+        ++iterator;
+    }
+    ++iterator; // is already '0'
+    for (size_t i = 0; i < s.size(); ++i)
+    {
+        *iterator = s[i] & 31;
+        ++iterator;
+    }
+    return result;
+}
+
+bool verify_checksum_base32(const std::string& hr_part, const data_chunk& data_part)
+{
+    data_chunk expanded = hrp_expand_base32(hr_part);
+    expanded.insert(expanded.end(), data_part.begin(), data_part.end());
+    return polymod_base32(expanded) == 1;
+}
+
+data_chunk create_checksum_base32(const std::string& hr_part, const data_chunk& data_part)
+{
+    data_chunk values = hrp_expand_base32(hr_part);
+    values.insert(values.begin(), data_part.begin(), data_part.end());
+    values.resize(values.size() + checksum_length);
+    uint32_t polymod = polymod_base32(values) ^ 1;
+    data_chunk result(checksum_length, 0);
+    for (size_t i = 0; i < result.size(); ++i)
+    {
+        result[i] = (polymod >> (5 * (5 - i))) & 31;
+    }
+    return result;
+}
+
+std::string encode_base32(const std::string& hr_part, const data_chunk& data_part)
+{
+    std::string encoded = hr_part + separator;
+    encoded.reserve(encoded.size() + data_part.size() + checksum_length);
+    data_chunk checksum = create_checksum_base32(hr_part, data_part);
+    for (auto value : data_part)
+    {
+        encoded += lut[value];
+    }
+    for (auto value : checksum)
+    {
+        encoded += lut[value];
+    }
+    return encoded;
+}
+
+inline char toLowercase(const char& c)
+{
+    return c + ('a' - 'A');
+}
+
+bool decode_base32(const std::string& in, std::string& hr_part, data_chunk& values)
+{
+    // Basic input check
+    if (in.length() < base32_min_length || in.length() > base32_max_length)
```

Comment:
Here, the check for the min length is redundant to the `pos_separator` checks on line 132. A string meeting the minimum length requirement still can be invalid.

Reply:
Yes this is true. I didn't write explicitly, but this first max length check assures that the method returns fast in case of a very large input string whilst find() needs to run over the whole string - worst case. So I suggest to keep this check.

Reply:
@marcoferiencik I agree that the max length check makes sense. I was talking about the min length check 😉

Reply:
Oh, I misread, of course this makes sense. Thanks!